### PR TITLE
added methods for refactoring

### LINKED ISSUE #35: Hardcode holding of jobs
The interaction between the event handler (`eessi_bot_software_layer.py`) and the job manager (`eessi_bot_job_manager.py`) is facilitated by the event handler submitting jobs with the parameter `--hold`, then the job manager releases jobs with the command `scontrol`. The parameter `--hold` is currently provided via the key `slurm_params` in the bot's `app.cfg`.

To avoid accidentally removing that parameter and thereby making the bot or interaction between its components dysfunctional, the parameter should be hardcoded for now.

### LINKED ISSUE #36: Refactor `build_easystack_from_pr` into smaller functions
The function `build_easystack_from_pr` in `tasks/build.py` has gotten quite long over time. Refactoring it into smaller functions should make the code more readable and also help reviewing future pull requests. Smaller functions might then also be reused by other components of the bot.

### PR DESCRIPTION
**Pull request by Hafsa-Naeem**
fixes #36, fixes #35

### CODE REVIEW

**Comment by Hafsa-Naeem** [comment]
fixes #36

File: tasks/build.py [hunk 1]
```diff
@@ -73,16 +74,17 @@ def build_easystack_from_pr(pr, event_info):
...
     # adopting approach outlined in https://github.com/EESSI/eessi-bot-software-layer/issues/17
     # need to use `base` instead of `head` ... don't need to know the branch name
     # TODO rename to base_repo_name?
     repo_name = pr.base.repo.full_name
     log("build_easystack_from_pr: pr.base.repo.full_name '%s'" % pr.base.repo.full_name)
     branch_name = pr.base.ref
     log("build_easystack_from_pr: pr.base.repo.ref '%s'" % pr.base.ref)
-
     jobs = []
     for arch_target,slurm_opt in arch_target_map.items():
```
Comment (boegel) [inline tasks/build.py:89]: suggestion: create functions for what's done *in* the loop body, something like:

```python
for arch_target,slurm_opt in arch_target_map.items():
        arch_job_dir = os.path.join(run_dir, arch_target.replace('/', '_'))
        mkdir(arch_job_dir)
        log("arch_job_dir '%s'" % arch_job_dir)

        set_up_pr_in(repo_name, pr_number, arch_job_dir)
        apply_customisations(arch_job_dir, cvmfs_customizations)

File: tasks/build.py [hunk 2]
```diff
@@ -161,12 +161,31 @@ def build_easystack_from_pr(pr, event_info):
...
+    return repo_name, jobs
+
 
+def build_easystack_from_pr(pr, event_info):
+    # retrieving some settings from 'app.cfg' in bot directory
+    # [github]
+    app_name = config.get_section('github').get('app_name')
+    
+    # [buildenv]
+    jobs_base_dir, local_tmp, build_job_script, submit_command, slurm_params, cvmfs_customizations, http_proxy, https_proxy, load_modules = get_build_env()
```
Comment (boegel) [inline tasks/build.py:177]: maybe this approach is better:

```python
build_env_cfg = get_build_env_configuration()
```
and use constants:
```python
SUBMIT_COMMAND = 'submit_command'
```
in `get_build_env_configuration`:
```python
    buildenv = config.get_section('buildenv')
    ...
    submit_command = buildenv.get(SUBMIT_COMMAND)
```

where you need the submit command, use:
```python
build_env_cfg[SUBMIT_COMMAND]
```
and pass down `build_env_cfg`:
```python
    repo_name, jobs = download_pull_request(pr, arch_target_map, run_dir, build_env_cfg)
    for job in jobs:
        submit_job(job, build_env_cfg)

**Review by boegel — CHANGES_REQUESTED** [review]

File: tasks/build.py [hunk 3]
```diff
@@ -52,16 +48,21 @@ def build_easystack_from_pr(pr, event_info):
...
-
-    # [architecturetargets]
+def get_architecturetargets():
     architecturetargets = config.get_section('architecturetargets')
     arch_target_map = json.loads(architecturetargets.get('arch_target_map'))
     log("arch target map '%s'" % json.dumps(arch_target_map))
+    return arch_target_map
+
 
+def create_directory(pr, jobs_base_dir, event_info):
```
Comment (boegel) [inline tasks/build.py:61]: Please also add a docstring, maybe also add [type hints](https://docs.python.org/3/library/typing.html) in the function signature:

```python
def create_directory(pr: int, jobs_base_dir: str, event_info: dict) -> (str, str, str)
    """
    Create directory for PR

    :param pr: PR number
    :param jobs_base_dir: location ...
    :returns: tuple of 3 elements: string with datestamp (<year>.<month>)
    """

**Review by boegel — CHANGES_REQUESTED** [review]

File: tasks/build.py [hunk 4]
```diff
@@ -101,138 +238,154 @@ def build_easystack_from_pr(pr, event_info):
         #
         #  - REPO_NAME is repo_name
         #  - PR_NUMBER is pr.number
```
Comment (boegel) [inline tasks/build.py:240]: move this comment into `setup_pr_in_arch_job_dir`?

File: tasks/build.py [hunk 5]
```diff
@@ -11,32 +11,58 @@
...
 
+JOBS_BASE_DIR = "jobs_base_dir"
+LOCAL_TMP = "local_tmp"
+BUILD_JOB_SCRIPT = "build_job_script"
+SUBMIT_COMMAND = "submit_command"
+SLURM_PARAMS = "slurm_params"
+CVMFS_CUSTOMIZATIONS = "cvmfs_customizations"
+HTTP_PROXY = "http_proxy"
+HTTPS_PROXY = "https_proxy"
+LOAD_MODULES = "load_modules"
```
Comment (boegel) [inline tasks/build.py:22]: please sort these alphabetically:

```python
BUILD_JOB_SCRIPT = ...
CVMFS_CUSTOMIZATIONS = ...
...
```

File: tasks/build.py [hunk 6]
```diff
@@ -11,32 +11,58 @@
...
+    
 
 def mkdir(path):
```
Comment (boegel) [inline tasks/build.py:25]: also docstring here?

File: tasks/build.py [hunk 7]
```diff
@@ -73,19 +131,98 @@ def build_easystack_from_pr(pr, event_info):
...
+    git_am_cmd = f'git am {pr.number}.patch'
+    log("Apply patch by running '%s' in directory '%s'" % (git_am_cmd,arch_job_dir))
+    patched = subprocess.run(git_am_cmd,
+                             cwd=arch_job_dir,
+                             shell=True,
+                             stdout=subprocess.PIPE,
+                             stderr=subprocess.PIPE)
+    log("Applied patch!\nStdout %s\nStderr: %s" % (patched.stdout,patched.stderr))
+
+def apply_cvmfs_customizations(cvmfs_customizations,arch_job_dir):
```
Comment (boegel) [inline tasks/build.py:186]: please add a docstring + two empty lines above function definition

File: tasks/build.py [hunk 8]
```diff
@@ -101,138 +238,154 @@ def build_easystack_from_pr(pr, event_info):
...
+
+    dt = datetime.now(timezone.utc)
+    job_comment += f'|{dt.strftime("%b %d %X %Z %Y")}|submitted|job waits for release by job manager|'
+
+    repo = gh.get_repo(repo_name)
+    pull_request = repo.get_pull(pr.number)
+    pull_request.create_issue_comment(job_comment)
+
 
+def build_easystack_from_pr(pr, event_info):
```
Comment (boegel) [inline tasks/build.py:360]: also add a docstring here

File: tasks/build.py [hunk 9]
```diff
@@ -11,32 +11,58 @@
...
+    config_data[BUILD_JOB_SCRIPT] = build_job_script
+
+    submit_command = buildenv.get(SUBMIT_COMMAND)
     log("submit_command '%s'" % submit_command)
-    slurm_params = buildenv.get('slurm_params')
+    config_data[SUBMIT_COMMAND] = submit_command
+    
+
+    #slurm_params = '--hold'
+    slurm_params = buildenv.get(SLURM_PARAMS)
```
Comment (boegel) [inline tasks/build.py:59]: This should become:

```python
slurm_params = buildenv.get(SLURM_PARAMS)
# always submit jobs with hold set, so job manager can release them
slurm_params += ' --hold'
```

which fixes #35

**Review by boegel — CHANGES_REQUESTED** [review]

**Review by boegel — APPROVED** [review]
lgtm, tested with https://github.com/boegel/software-layer/pull/11 :+1:

**Pull request merged**